probe-rs/src/session.rs

```diff
@@ -112,260 +118,296 @@ impl ArchitectureInterface {
+    /// steps taken are detailed below for ARM and RISCV.
+    ///
+    /// ## ARM
+    ///
+    /// - The actual target is determined using the given TargetSelector`.
+    ///
+    /// - If attach under reset is selected, the [`ArmDebugSequence::reset_hardware_assert()`] function is called,
+    ///   if the probe is CMSIS-DAP probe or a J-Link. For other probes, where a custom sequence is not supported,
+    ///   the [`Probe::target_reset_assert()`](crate::Probe::target_reset_assert()) function is called.
+    ///
+    ///
+    pub(crate) fn new(probe: Probe, target: TargetSelector, config: Config) -> Result<Self, Error> {
+        let (probe, target) = get_target_from_selector(target, config.attach_method, probe)?;
 
         let mut session = match target.architecture() {
-            Architecture::Arm => {
-                let config = target.cores[0].clone();
-                let arm_core_access_options = match config.core_access_options {
-                    probe_rs_target::CoreAccessOptions::Arm(opt) => opt,
-                    probe_rs_target::CoreAccessOptions::Riscv(_) => {
-                        unreachable!("This should never happen. Please file a bug if it does.")
-                    }
-                };
+            Architecture::Arm => Session::attach_arm(probe, target, config)?,
+            Architecture::Riscv => Session::attach_riscv(target, probe)?,
+        };
 
-                let default_memory_ap = MemoryAp::new(ApAddress {
-                    dp: match arm_core_access_options.psel {
-                        0 => DpAddress::Default,
-                        x => DpAddress::Multidrop(x),
-                    },
-                    ap: arm_core_access_options.ap,
-                });
+        session.clear_all_hw_breakpoints()?;
 
-                let sequence_handle = match &target.debug_sequence {
-                    DebugSequence::Arm(sequence) => sequence.clone(),
-                    DebugSequence::Riscv(_) => {
-                        panic!("Mismatch between architecture and sequence type!")
-                    }
-                };
+        Ok(session)
+    }
 
-                if AttachMethod::UnderReset == attach_method {
-                    let span = tracing::debug_span!("Asserting hardware assert");
-                    let _enter = span.enter();
-
-                    if let Some(dap_probe) = probe.try_as_dap_probe() {
-                        sequence_handle.reset_hardware_assert(dap_probe)?;
-                    } else {
-                        tracing::info!(
-                            "Custom reset sequences are not supported on {}.",
-                            probe.get_name()
-                        );
-                        tracing::info!("Falling back to standard probe reset.");
-                        probe.target_reset_assert()?;
-                    }
+    /// Automatically creates a session with the first connected probe found.
+    #[tracing::instrument(skip(target))]
```

Comment:
How can we make this properly print? Because I think the target would be very useful info. Maybe RN it's mostly known but auto attach would profit off it.